libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -451,6 +478,144 @@ uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
+                 << reward_votes_pbft_block_hash_;
+    assert(false);
+  }
+
+  if (!verifyRewardVote(vote)) return false;
+
+  reward_votes_.insert({vote_hash, vote});
+  db_->saveLastBlockCertVote(vote);
+  LOG(log_nf_) << "add reward vote " << vote_hash;
+
+  return true;
+}
+
+bool VoteManager::verifyRewardVote(const std::shared_ptr<Vote>& vote) {
+  if (vote->getType() != cert_vote_type) {
+    LOG(log_er_) << "Reward vote type is not cert vote type " << *vote;
+    return false;
+  }
+
+  if (vote->getStep() != 3) {
+    LOG(log_er_) << "Reward vote step is not 3 " << *vote;
+    return false;
+  }
+
+  const uint64_t dpos_period = pbft_chain_->getPbftChainSize() - 1;  // reward period - 1
+  const auto& voter_account_addr = vote->getVoterAddr();
+  uint64_t voter_dpos_votes_count;
+  try {
+    voter_dpos_votes_count = final_chain_->dpos_eligible_vote_count(dpos_period, voter_account_addr);
+  } catch (state_api::ErrFutureBlock& c) {
+    LOG(log_er_) << c.what() << ". Voter account " << voter_account_addr << " in DPOS period<" << dpos_period
+                 << "> is too far for ahead of DPOS. Have executed chain size " << final_chain_->last_block_number();
+    return false;
+  }
+
+  if (!voter_dpos_votes_count) {
+    LOG(log_er_) << "Invalid reward vote: voter was not eligible to vote " << *vote;
+    return false;
+  }
+
+  const auto dpos_total_votes_count = final_chain_->dpos_eligible_total_vote_count(dpos_period);
+  const size_t pbft_sortition_threshold = std::min<size_t>(pbft_committee_size_, dpos_total_votes_count);
+  try {
+    vote->validate(voter_dpos_votes_count, dpos_total_votes_count, pbft_sortition_threshold);
+  } catch (const std::logic_error& e) {
+    LOG(log_er_) << e.what();
+    LOG(log_er_) << "DPOS period " << dpos_period << ", voter " << voter_account_addr << " votes count "
+                 << voter_dpos_votes_count << ", total votes count " << dpos_total_votes_count
+                 << ", pbft sortition threshold " << pbft_sortition_threshold;
+
+    return false;
+  }
+
+  return true;
+}
+
+bool VoteManager::checkRewardVotes(const std::shared_ptr<PbftBlock>& pbft_block) {
+  if (pbft_block->getPeriod() == 1) {
+    // First period no reward votes
+    return true;
+  }
+
+  std::vector<vote_hash_t> missing_reward_votes;
+  const auto& reward_votes_hashes = pbft_block->getRewardVotes();
+  {
+    std::shared_lock lock(reward_votes_mutex_);
+    for (const auto& v : reward_votes_hashes) {
+      if (!reward_votes_.contains(v)) {
+        missing_reward_votes.emplace_back(v);
+      }
+    }
+  }
+
+  if (!missing_reward_votes.empty()) {
+    std::ostringstream missing_reward_votes_log;
+    missing_reward_votes_log << "Missing reward votes: ";
+    for (const auto& v : missing_reward_votes) {
+      missing_reward_votes_log << " " << v.toString();
+    }
+    LOG(log_er_) << missing_reward_votes_log.str();
+
+    return false;
+  }
```

Comment:
as I commented before, no need to get all missing votes, just one should be enough

Reply:
FIxed